File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = left + ((right - left) // 2)
            print("nums[mid] is", nums[mid])
            # now, we check whether target is larger, smaller or equal to mid to determine where to shift our pointers
            if nums[mid] == target:
                return mid
            if nums[left] <= nums[mid]:
                # This means the left half is perfectly sorted
                # Hence, we can check here in this perfectly sorted array whether target lies in this region
                if nums[left] <= target and target <= nums[mid]:
                    right = mid - 1
                else:
                    # This means it is on the other side
                    left = mid + 1
            else:
                # This means the right half is perfectly sorted
                if nums[mid] <= target and target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        return -1

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.java
class Solution {
    public int findMin(int[] nums) {
        int left = 0;
        int right = nums.length - 1;
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            // In a rotated sorted array, only one half of the array can be possibly out of order
            // 1. the right side is out of order
            if (nums[mid] > nums[right]) {
                // This means the array fell on the right side, which means that smallest element lies on the right side of the array for sure
                left = mid + 1;
            } else {
                // if nums[mid] <= nums[right], this means the minimum element lies on the left OR is on mid itself
                right = mid;
            }

        }
        // Both pointers will eventually converge 
        return nums[left];

    }
}
